Source/DafnyRuntime/DafnyRuntime.py

```diff
@@ -39,57 +45,159 @@ def _break(name):
 def _tail_call():
     raise TailCall()
 
-class Seq(list):
+class Seq(tuple):
     def __init__(self, __iterable = None, isStr = False):
         if __iterable is None:
             __iterable = []
-        self.isStr = isinstance(__iterable, str) or isStr
-        super().__init__(__iterable)
+        self.isStr = isStr or isinstance(__iterable, str) or (isinstance(__iterable, Seq) and __iterable.isStr) or (not isinstance(__iterable, GeneratorType) and all(isinstance(e, str) and len(e) == 1 for e in __iterable) and len(__iterable) > 0)
 
     @property
     def Elements(self):
         return self
 
     @property
     def UniqueElements(self):
-        return set(self)
+        return frozenset(self)
 
     def __repr__(self) -> str:
         if self.isStr:
-          return ''.join(self)
-        return super().__repr__()
+            return ''.join(self)
+        return '[' + ', '.join(map(repr, self)) + ']'
 
     def __add__(self, other):
         return Seq(super().__add__(other), isStr=self.isStr and other.isStr)
 
+    def __getitem__(self, key):
+        if isinstance(key, slice):
+            indices = range(*key.indices(len(self)))
+            return Seq((self[i] for i in indices), isStr=self.isStr)
+        return super().__getitem__(key)
+
+    def set(self, key, value):
+        l = list(self)
+        l[key] = value
+        return Seq(l, isStr=self.isStr)
+
     def __hash__(self) -> int:
         return hash(tuple(self))
 
-class Set(set):
+class Set(frozenset):
     @property
     def Elements(self):
         return self
 
+    @property
+    def AllSubsets(self):
+        # https://docs.python.org/3/library/itertools.html#itertools-recipes
+        s = list(self)
+        return map(Set, chain.from_iterable(combinations(s, r) for r in range(len(s)+1)))
+
     def __str__(self) -> str:
         return '{' + ', '.join(map(str, self)) + '}'
 
     def union(self, other):
-        return Set(set.union(self, other))
+        return Set(super().union(self, other))
+
+    def intersection(self, other):
+        return Set(super().intersection(other))
+
+    def ispropersubset(self, other):
+        return self.issubset(other) and self != other
 
     def __or__(self, other):
         return self.union(other)
 
     def __sub__(self, other):
         return Set(super().__sub__(other))
 
+class MultiSet(Counter):
+    def __str__(self) -> str:
+        return 'multiset{' + ', '.join(map(str, self.elements())) + '}'
+
+    def __len__(self):
+        return reduce(lambda acc, key: acc + self[key], self, 0)
+
+    def union(self, other):
+        return MultiSet(self + other)
+
+    def __or__(self, other):
+        return self.union(other)
+
+    def intersection(self, other):
+        return MultiSet(self & other)
+
+    def __sub__(self, other):
+        return MultiSet(super().__sub__(other))
+
+    @property
+    def keys(self):
+        return (key for key in self if self[key] > 0)
```

Comment:
You need to override `__eq__` and `__hash__` for this class as well, since `Counter` doesn't consider a missing key equivalent to one mapped to `0`:

```python
>>> c = Counter()
>>> c[4] = 0
>>> c2 = Counter()
>>> c == c2
False
```

Reply:
`hash` is fine, as it uses keys, but I'll look into updating `__eq__`.